### PR TITLE
[IOHK-467] Add transaction tracers to ChainDB for tracing how long does it take for a tx to be integrated into a block.

### PR DESCRIPTION
**Pull request by ksaric**
https://github.com/input-output-hk/iohk-monitoring-framework/issues/467

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs [hunk 1]
```diff
@@ -281,9 +281,12 @@ implAddTxs mpEnv accum txs = assert (all txInvariant txs) $ do
...
 
-    traceBatch TraceMempoolRemoveTxs   mempoolSize (map fst removed)
-    traceBatch TraceMempoolAddTxs      mempoolSize accepted
-    traceBatch TraceMempoolRejectedTxs mempoolSize rejected
+    -- We get the time for the measurement of the txs added into the mempool.
+    time <- getMonotonicTime
+
+    traceBatch TraceMempoolRemoveTxs   mempoolSize (map fst removed) time
+    traceBatch TraceMempoolAddTxs      mempoolSize accepted time
+    traceBatch TraceMempoolRejectedTxs mempoolSize rejected time
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs]: Can you align on `time`? We like alignment, as it helps us spot the differences between similar looking lines.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs [hunk 2]
```diff
@@ -1,4 +1,6 @@
 {-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE StandaloneDeriving #-}
+
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs]: Can you align the `#-}`? We use `stylish-haskell` which does this for us. The next we touch this file, we'd automatically make this change.

Can you can remove the new empty line below as well?

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs [hunk 3]
```diff
@@ -97,7 +100,7 @@ showTracers tr = Tracers
 -------------------------------------------------------------------------------}
 
 -- | Trace the forging of a block as a slot leader.
-data TraceForgeEvent blk
+data TraceForgeEvent blk txs
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs]: Remove the `txs` parameter, we know the type of the transactions: `[GenTx blk]`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs [hunk 4]
```diff
@@ -123,3 +127,4 @@ data TraceForgeEvent blk
   -- validation and the ledger validation. This is a serious error!
   | TraceForgedInvalidBlock SlotNo blk (InvalidBlockReason blk)
   deriving (Eq, Show)
+
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs]: Unnecessary diff

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs [hunk 5]
```diff
@@ -361,8 +361,13 @@ forkBlockProduction IS{..} =
           ChainDB.addBlock chainDB newBlock
           -- Check whether we adopted our block
           curTip <- atomically $ ChainDB.getTipPoint chainDB
+          -- We measure the time directly.
+          -- We could also measure it when collecting the trace events, but this
+          -- is much more precise.
+          time <- getMonotonicTime
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: I would move this to the `then` branch, because we only use it there. It doesn't really matter much, though.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs (hunk 2 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs]: Ah, you removed. This is no longer relevant.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs (hunk 3 above)
Comment (ksaric) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs]: And how do I derive the `Eq` and `Show` for that?
Isn't that still abstract at this point and it is hard to derive?

Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs]:
```haskell
deriving instance (ProtocolLedgerView blk, Eq blk, Eq (GenTx blk))
               => Eq   (TraceForgeEvent blk)
deriving instance (ProtocolLedgerView blk, Show blk, Show (GenTx blk))
               => Show (TraceForgeEvent blk)
```
That requires `StandaloneDeriving` and `UndecidableInstances`.

*However,* I realise now that there is some precedent in the tracers (`txInboundTracer`/`txOutboundTracer`) already. Can you just change the type parameter to `tx` instead of `txs`? You can wrap `tx` in a list inside of `TraceForgeEvent`.

Comment (ksaric) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs]: Will fix.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs [hunk 6]
```diff
@@ -1,4 +1,5 @@
 {-# LANGUAGE FlexibleContexts #-}
+
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs]: Eh, can I ask you to remove this empty line too?

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs [hunk 7]
```diff
@@ -283,19 +283,25 @@ data TraceEventMempool blk
       -- ^ New, valid transaction were added to the Mempool.
       !Word
       -- ^ The total number of transactions now in the Mempool.
+      !Time
+      -- ^ The time at which the transaction were added.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs]:
```suggestion
      -- ^ The time at which the transactions were added.
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs [hunk 8]
```diff
@@ -283,19 +283,25 @@ data TraceEventMempool blk
...
   | TraceMempoolRejectedTxs
       ![(GenTx blk, ApplyTxErr blk)]
       -- ^ New, invalid transaction were rejected and thus not added to the
       -- Mempool.
       !Word
       -- ^ The total number of transactions now in the Mempool.
+      !Time
+      -- ^ The time at which the transaction were rejected.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs]:
```suggestion
      -- ^ The time at which the transactions were rejected.
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs [hunk 9]
```diff
@@ -283,19 +283,25 @@ data TraceEventMempool blk
...
   | TraceMempoolRemoveTxs
       ![GenTx blk]
       -- ^ Previously valid transactions that are no longer valid because of
       -- changes in the ledger state. These transactions have been removed
       -- from the Mempool.
       !Word
       -- ^ The total number of transactions now in the Mempool.
+      !Time
+      -- ^ The time at which the transaction were removed.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs]:
```suggestion
      -- ^ The time at which the transactions were removed.
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs [hunk 10]
```diff
@@ -281,9 +281,12 @@ implAddTxs mpEnv accum txs = assert (all txInvariant txs) $ do
 
     let ValidationResult { vrNewValid = accepted } = vr
 
-    traceBatch TraceMempoolRemoveTxs   mempoolSize (map fst removed)
-    traceBatch TraceMempoolAddTxs      mempoolSize accepted
-    traceBatch TraceMempoolRejectedTxs mempoolSize rejected
+    -- We get the time for the measurement of the txs added into the mempool.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs]:
```suggestion
    -- We record the time at which the transactions were added to the mempool so we can use it in our performance measurements
```

I would also be okay with removing this comment altogether.

**Comment by ksaric** [comment]
:tada:
